### PR TITLE
Drop do while

### PR DESCRIPTION
**Pull request by odersky**
Drop
```
 do <body> while <cond>
```
in favor of
```
  while ({ <body> ; <cond> }) ()
```
Under -rewrite one syntax can be rewritten to the other.

Reasons:

 - do-while is used relatively rarely and it can expressed faithfully using just `while`. So there seems to be little point in having it as a separate syntax construct.
 - we'd like to reclaim the `do` as a statement continuation token.
 - Using `while` as the only loop syntax gives an elegant solution to the "loop-and-a-half" problem. E.g.
```scala
var x: Int = 0
while ({
  x = iterator.next
  x >= 0
}) println(x)
```

### CODE REVIEW

**Comment by propensive** [comment]
It seems odd, initially, not to have the statement and predicate wrapped in `{}`, but (in your example), the braces would also imply a constrained scope of `x`, so I can see a lot of sense in this, and it's consistent with the bracketless `for` we now have.

I checked to see if the predicate in an `if`/`then`/`else` could also facilitate a statement, and it can't, though for consistency (not so much for compelling utility), I think it ought to, e.g.
```
def foo(x: Iterator[Int]) = if
  val h = x.next
  h >= 0
then s"+$h" else s"-$h"
```
This would help to establish an invariant that new identifiers introduced in expressions comprising of multiple keywords would be scoped to the entire expression.

However, would this work? There's an argument to be made (though I'm not sure I have strong opinions) that a block which evaluates to a `Boolean` used in an expression position should not have it's scope leak outside the limits of that block.
```
while {
  val x = iterator.next
  x >= 0
} do println(x)
```

Another quirk of the new syntax is that a multi-line predicate can be written without braces, whereas a multi-line action cannot, though it seems like the latter would be by far the more common case.

**Comment by odersky** [comment]
@propensive The brace-less syntax in `while` slipped in by accident. That's not implemented (yet?) Details remain to be worked out. For the moment I have put the braces back.

**Comment by propensive** [comment]
> @propensive The brace-less syntax in `while` slipped in by accident. That's not implemented (yet?)

Ok, great. I'll wait and see. :)

File: compiler/src/dotty/tools/dotc/util/CommentParsing.scala [hunk 1]
```diff
@@ -172,8 +172,10 @@ object CommentParsing {
   def skipVariable(str: String, start: Int): Int = {
     var idx = start
     if (idx < str.length && (str charAt idx) == '{') {
-      do idx += 1
-      while (idx < str.length && (str charAt idx) != '}')
+      while {
+        idx += 1
+        idx < str.length && (str charAt idx) != '}'
+      } do ()
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/util/CommentParsing.scala]: For this PR we should only remove `do .. while` but not add `while .. do`

Comment (odersky) [reply compiler/src/dotty/tools/dotc/util/CommentParsing.scala]: It's actually already supported by the existing code. I converted to a mixture of the two to test both forms.

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 2]
```diff
@@ -1339,12 +1330,30 @@ object Parsers {
           WhileDo(cond, body)
         }
       case DO =>
-        atSpan(in.skipToken()) {
+        in.errorOrMigrationWarning(
+          i"""`do <body> while <cond>' is no longer supported,
+             |use `while ({<body> ; <cond>}) ()' instead.
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]:
```suggestion
             |use `while {<body> ; <cond>} do ()' instead.
```

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 3]
```diff
@@ -1339,12 +1330,30 @@ object Parsers {
...
+            patch(source, Span(whileStart, whileStart + 5), ";")
+            cond match {
+              case Parens(_) =>
+                patch(source, Span(cond.span.start, cond.span.start + 1), "")
+                patch(source, Span(cond.span.end - 1, cond.span.end), "")
+              case _ =>
+            }
+            patch(source, cond.span.endPos, "}) ()")
+          }
+          WhileDo(Block(Block(Nil, body), Block(Nil, cond)), Literal(Constant(())))
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]:
```suggestion
          WhileDo(Block(body :: Nil, cond), Literal(Constant(())))
```

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 4]
```diff
@@ -1339,12 +1330,30 @@ object Parsers {
           WhileDo(cond, body)
         }
       case DO =>
-        atSpan(in.skipToken()) {
+        in.errorOrMigrationWarning(
+          i"""`do <body> while <cond>' is no longer supported,
+             |use `while {<body> ; <cond>} do ()' instead.
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: I'm pretty sure we discussed today that this PR shouldn't be mixed up with that possible future syntax. Hence I believe we should suggest `while ({ body; cond }) ()` here instead.

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 5]
```diff
@@ -1339,12 +1330,30 @@ object Parsers {
...
+            patch(source, Span(whileStart, whileStart + 5), ";")
+            cond match {
+              case Parens(_) =>
+                patch(source, Span(cond.span.start, cond.span.start + 1), "")
+                patch(source, Span(cond.span.end - 1, cond.span.end), "")
+              case _ =>
+            }
+            patch(source, cond.span.endPos, "}) ()")
+          }
+          WhileDo(Block(body :: Nil, cond), Literal(Constant(())))
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: This is incorrect, as both `body` and `block` could define symbols with the same name. If they are put together in a single `Block`, the scopes will clash.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: We will need a test for this.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: Actually, `body` and `cond` are already blocks if they have definitions. I added some tests in #7002. Did I miss some other definitions that could be represented differently?

**Pull request merged**